Make client_uid variable a flexible array member of client structure - and allocate / deallocate together

mykaul opened the pull request:
1. Similar to multiple previous PRs, it makes it part of the client structure. It also removes the unneeded counter - it is a fixed size.
2. Made scratch ctx a fixed array (as it actually was), allocated and deallocated as part of the client struct.

(note - client_uid could also be a flexible array member in other structs, such as lock_migration_info_t)

Commit:
client_t: move scratch structure variables to be directly accessible from the client struct

Easier to read and a step into allocating with the client struct itself.

Updates: #1000
Signed-off-by: Yaniv Kaul <ykaul@redhat.com>

Commit:
client_t: remove not needed scratch ctx count

It doesn't change - it's always GF_CLIENTCTX_INITIAL_SIZE

Updates: #1000
Signed-off-by: Yaniv Kaul <ykaul@redhat.com>

Commit:
client_t: turn client scratch ctx to a flexible array member variable of client struct

Allocate it along with the whole client structure allocation.
Useful, as it's fixed size, the allocation is done under lock, and for locality.

Updates: #1000
Signed-off-by: Yaniv Kaul <ykaul@redhat.com>

Commit:
mem-types.h: remove unused memory type

The code is under #ifdef, commented out at the moment.

Updates: #1000
Signed-off-by: Yaniv Kaul <ykaul@redhat.com>

gluster-ant commented:
CLANG-FORMAT FAILURE:
Before merging the patch, this diff needs to be considered for passing clang-format

```diff --git a/libglusterfs/src/glusterfs/mem-types.h b/libglusterfs/src/glusterfs/mem-types.h
index e9ac1d6a9..85cc3cb93 100644
--- a/libglusterfs/src/glusterfs/mem-types.h
+++ b/libglusterfs/src/glusterfs/mem-types.h
@@ -75,10 +75,11 @@ enum gf_common_mem_types_ {
     gf_common_mt_cliententry_t, /* used only in one location */
     gf_common_mt_clienttable_t, /* used only in one location */
     gf_common_mt_client_t,      /* used only in one location */
-    //gf_common_mt_client_ctx,    /* unused - could either be removed or code needs to be undef */
-    gf_common_mt_auxgids,       /* used only in one location */
-    gf_common_mt_syncopctx,     /* used only in one location */
-    gf_common_mt_iobrefs,       /* used only in one location */
+    // gf_common_mt_client_ctx,    /* unused - could either be removed or code
+    // needs to be undef */
+    gf_common_mt_auxgids,   /* used only in one location */
+    gf_common_mt_syncopctx, /* used only in one location */
+    gf_common_mt_iobrefs,   /* used only in one location */
     gf_common_mt_gsync_status_t,
     gf_common_mt_uuid_t,
     gf_common_mt_mgmt_v3_lock_obj_t, /* used only in one location */
```

mykaul commented:
/run regression

Commit:
client_t.c: move client_uid to be a flexible array member (insted of scratch ctx, which is actually constant size)

It wasn't the best choice to move scratch ctx to be a variable size - it isn't - it's very fixed and constant.
So turned client_uid into such.

Updates: #1000
Signed-off-by: Yaniv Kaul <ykaul@redhat.com>

gluster-ant commented:
CLANG-FORMAT FAILURE:
Before merging the patch, this diff needs to be considered for passing clang-format

```diff --git a/libglusterfs/src/glusterfs/mem-types.h b/libglusterfs/src/glusterfs/mem-types.h
index e9ac1d6a9..85cc3cb93 100644
--- a/libglusterfs/src/glusterfs/mem-types.h
+++ b/libglusterfs/src/glusterfs/mem-types.h
@@ -75,10 +75,11 @@ enum gf_common_mem_types_ {
     gf_common_mt_cliententry_t, /* used only in one location */
     gf_common_mt_clienttable_t, /* used only in one location */
     gf_common_mt_client_t,      /* used only in one location */
-    //gf_common_mt_client_ctx,    /* unused - could either be removed or code needs to be undef */
-    gf_common_mt_auxgids,       /* used only in one location */
-    gf_common_mt_syncopctx,     /* used only in one location */
-    gf_common_mt_iobrefs,       /* used only in one location */
+    // gf_common_mt_client_ctx,    /* unused - could either be removed or code
+    // needs to be undef */
+    gf_common_mt_auxgids,   /* used only in one location */
+    gf_common_mt_syncopctx, /* used only in one location */
+    gf_common_mt_iobrefs,   /* used only in one location */
     gf_common_mt_gsync_status_t,
     gf_common_mt_uuid_t,
     gf_common_mt_mgmt_v3_lock_obj_t, /* used only in one location */
```

mykaul commented:
/run regression

gluster-ant commented:
0 test(s) failed 


1 test(s) generated core 
./tests/000-flaky/basic_afr_split-brain-favorite-child-policy.t

1 test(s) needed retry 
./tests/000-flaky/basic_afr_split-brain-favorite-child-policy.t

1 flaky test(s) marked as success even though they failed 
./tests/000-flaky/basic_afr_split-brain-favorite-child-policy.t
https://build.gluster.org/job/gh_centos7-regression/3013/

Commit:
Remove additional GF_FREE() that is not needed.

gluster-ant commented:
CLANG-FORMAT FAILURE:
Before merging the patch, this diff needs to be considered for passing clang-format

```diff --git a/libglusterfs/src/glusterfs/mem-types.h b/libglusterfs/src/glusterfs/mem-types.h
index e9ac1d6a9..85cc3cb93 100644
--- a/libglusterfs/src/glusterfs/mem-types.h
+++ b/libglusterfs/src/glusterfs/mem-types.h
@@ -75,10 +75,11 @@ enum gf_common_mem_types_ {
     gf_common_mt_cliententry_t, /* used only in one location */
     gf_common_mt_clienttable_t, /* used only in one location */
     gf_common_mt_client_t,      /* used only in one location */
-    //gf_common_mt_client_ctx,    /* unused - could either be removed or code needs to be undef */
-    gf_common_mt_auxgids,       /* used only in one location */
-    gf_common_mt_syncopctx,     /* used only in one location */
-    gf_common_mt_iobrefs,       /* used only in one location */
+    // gf_common_mt_client_ctx,    /* unused - could either be removed or code
+    // needs to be undef */
+    gf_common_mt_auxgids,   /* used only in one location */
+    gf_common_mt_syncopctx, /* used only in one location */
+    gf_common_mt_iobrefs,   /* used only in one location */
     gf_common_mt_gsync_status_t,
     gf_common_mt_uuid_t,
     gf_common_mt_mgmt_v3_lock_obj_t, /* used only in one location */
```

mykaul commented:
/run regression

xhernandez commented on libglusterfs/src/glusterfs/mem-types.h:
```diff
@@ -75,7 +75,7 @@ enum gf_common_mem_types_ {
     gf_common_mt_clienttable_t, /* used only in one location */
     gf_common_mt_client_t,      /* used only in one location */
-    gf_common_mt_client_ctx,    /* used only in one location */
+    //gf_common_mt_client_ctx,    /* unused - could either be removed or code needs to be undef */
```
It can be removed.

mykaul replied:
done.

Commit:
Fix review comments - remove unused memory type

mykaul commented:
/run regression
